pkg/cli/devicetree_test.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2019-present Open Networking Foundation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cli
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/onosproject/onos-config/pkg/northbound/admin"
+	"github.com/onosproject/onos-config/pkg/northbound/diags"
+	"gotest.tools/assert"
+	"io"
+	"regexp"
+	"strings"
+	"testing"
+	"time"
+)
+
+var configRecvCount = 0
+
+func diagConfigRecv() (*diags.Configuration, error) {
+	if configRecvCount == 0 {
+		now := time.Now()
+		result := &diags.Configuration{
+			Name:       "config1",
+			DeviceID:   "device1",
+			Version:    "1.0.0",
+			DeviceType: "switch",
+			Created:    &now,
+			Updated:    &now,
+			ChangeIDs:  []string{"2Lo0ZC0wqhuwbnSF9px6kDvEEWI="},
+		}
+		configRecvCount++
+		return result, nil
+	}
+	return nil, io.EOF
+}
+
+var changesRecvCount = 0
+
+func diagChangesRecv() (*admin.Change, error) {
+	if changesRecvCount < 3 {
+		now := time.Now()
+		result := &admin.Change{
+			Time: &now,
+			Id:   "2Lo0ZC0wqhuwbnSF9px6kDvEEWI=",
+			Desc: "Change1",
+			ChangeValues: []*admin.ChangeValue{
+				{Path: "/a/b/c",
```

Comment:
To create a ChangeValue you could use change.NewChangeValue(path, value, isRemove) here, and create value with change.NewTypedValueString("VALUE")

Then with the ChangeValue(s) you could use change.NewChange and put that straight in to `result` - it will generate a hash - but that will stay consistent

Reply:
These changes would suffer from an impedance mismatch between admin.Change and store.Change. I would have to convert from the store based change to an admin based change, creating two different versions of the change. Additionally, this test is not testing store functionality, and actually has no store in it. All it is testing is that the formatting of the data is correct. That data itself has no meaning. I don't think the extra code to generate the data using the utility routines is necessary. We'd be replacing 10 lines of code to create the object directly with at least 25 lines of code to invoke the store routines. I would prefer not to do this.